src/testing/slim_check/sampleable.lean

```diff
@@ -48,134 +64,179 @@ namespace slim_check
 functions between the two, going in both directions.
+
+Function `g` is used to define the well-founded order that
+`shrink` is expected to follow.
 -/
-def sampleable.lift (α : Type u) {β : Type u} [sampleable α] (f : α → β) (g : β → α) : sampleable β :=
-{ sample := f <$> sample α,
-  shrink := λ x, f <$> shrink (g x) }
+def sampleable.lift (α : Type u) {β : Type u} [sampleable α] (f : α → β) (g : β → α)
+  (h : ∀ (a : α), sizeof (g (f a)) ≤ sizeof a) : sampleable β :=
+{ wf := ⟨ sizeof ∘ g ⟩,
+  sample := f <$> sample α,
+  shrink := λ x,
+    have ∀ a,  sizeof a < sizeof (g x) → sizeof (g (f a)) < sizeof (g x),
+      by introv h'; solve_by_elim [lt_of_le_of_lt],
+    subtype.map f this <$> shrink (g x) }
 
 instance nat.sampleable : sampleable ℕ :=
 { sample := sized $ λ sz, coe <$> choose_any (fin $ succ (sz^3)) <|>
                           coe <$> choose_any (fin $ succ sz),
-  shrink := lazy_list.of_list ∘ nat.shrink }
+  shrink :=  λ x, lazy_list.of_list $ nat.shrink x }
+
+/-- `iterate_shrink p x` takes a decidable predicate `p` and a
+value `x` of some sampleable type and recursively shrinks `x`.
+It first calls `shrink x` to get a list of candidate sample,
+finds the first that satisfies `p` and recursively tries
+to shrink that one. -/
+def iterate_shrink {α} [has_to_string α] [sampleable α]
+  (p : α → Prop) [decidable_pred p] :
+  α → option α :=
+well_founded.fix has_well_founded.wf $ λ x f_rec,
+  do trace sformat!"{x} : {(shrink x).to_list}" $ pure (),
+     y ← (shrink x).find (λ a, p a),
+     f_rec y y.property <|> some y.val .
 
 instance sampleable_fin {n} [fact $ 0 < n] : sampleable (fin n) :=
-sampleable.lift ℕ (fin.of_nat') subtype.val
+sampleable.lift ℕ fin.of_nat' subtype.val $
+λ i, (mod_le _ _ : i % n ≤ i)
 
 @[priority 100]
 instance sampleable_fin' {n} : sampleable (fin (succ n)) :=
-sampleable.lift ℕ fin.of_nat subtype.val
+sampleable.lift ℕ fin.of_nat subtype.val $
+λ i, (mod_le _ _ : i % succ n ≤ i)
 
-instance sampleable_pnat : sampleable ℕ+ :=
-sampleable.lift ℕ nat.succ_pnat (λ i, i - 1)
-
-/-- `int.shrink' k n` creates a list of integers by successively
-dividing `n` by 2, subtracting the result from `k` and alternating the signs.
-For example, `int.shrink 40 = [20, -20, 30, -30, 35, -35, 38, -38, 39, -39]`. -/
-def int.shrink' (k : ℕ) : ℕ → list ℤ → list ℤ
-| n ls :=
-if h : 1 < n
-  then
-    have n / 2 < n, from div_lt_self (by linarith) (by norm_num),
-    let m := n / 2 in
-    let m' := k - m in
-    int.shrink' m (-↑m' :: m' :: ls)
-  else ls.reverse
+/-- Predecessor of a `ℕ+`, as a `ℕ`. -/
+def pnat.pred_nat (i : ℕ+) : ℕ := i - 1
 
-/-- `int.shrink n` creates a list of integers by successively
-dividing by 2, subtracting the result from `n` and alternating the signs.
-For example, `int.shrink 40 = [20, -20, 30, -30, 35, -35, 38, -38, 39, -39]`. -/
-def int.shrink (i : ℤ) : list ℤ :=
-int.shrink' (int.nat_abs i) (int.nat_abs i) []
+instance sampleable_pnat : sampleable ℕ+ :=
```

Comment:
```suggestion
instance pnat.sampleable : sampleable ℕ+ :=
```
and similar changes elsewhere, or change others to this format. Some are `type.sampleable` and others are `sampleable_type`